### DISCUSSION TITLE
Difference between createMachine and Machine

### DISCUSSION THREAD

**Discussion by shinenic** [statelyai/xstate#1891] [Q&A] [answered]
Hello guys,
I was confused about what's the difference between `createMachine` and `Machine`.
I see both of them are used in the following example
https://codesandbox.io/s/xstate-react-toast-example-ri36g?file=/src/index.js
but I have no idea when should I use `createMachine` or `Machine`.

**Comment by Andarist** [comment, marked as answer]
Both are pretty much the same when in terms of runtime characteristics - they only differ in what TS generics they accept (mainly the `createMachine` doesn't accept the `TSchema` generic)

**Reply by davidkpiano** [reply]
Right, and the way forward would be to use `createMachine` in v5.

**Reply by joffarex** [reply]
I was wondering, would there be no way of explicitly defining `TStateSchema` on `createMachine`? Right now it to `StateMachine` type, from `createMachine`, that parameter is passed as `any`. I am not using `Typegen`, I have explicitly typed everything.

**Reply by Andarist** [reply]
There is no way to use `TStateSchema` with `createMachine`. This generic is being deprecated and going to be removed in v5. It's derivative information from the used machine and we don't believe it provides much value. Could you tell us more about your use case? Why do you want to use it?

**Reply by joffarex** [reply]
I simply wanted to have type safety while checking for states outside machine with, for example, `states.matches` without using typegen.

Currently without typegen, you can't really do that other than using enums/constant variables in both places, but that's not really a type safety.

**Reply by Andarist** [reply]
What is stopping you from adopting the typegen?

**Reply by joffarex** [reply]
In my use-case, we already had some kind of "state machine" in the project, hand-made basically, from which we have lots of enum left-overs and did not want to do "magic strings" yet.

**Reply by Andarist** [reply]
Hm, it seems that the only "magic strings" you'd have to introduce with typegen would be at the `createMachine` call site. The consuming sited could still use your constants - if they would only match the declared state schema. That would, of course, disassociate those constants from their "source of truth" but at least you wouldn't have to refactor a lot of the files at once. How do you feel about it?

**Reply by joffarex** [reply]
Unfortunately I would not really like disconnecting state names in the machine from the ones that we'd use during state machine "consumption". We have lots of people working on the project and it could get really messy really quickly having to keep that. 

I think we can live right now with using enums and not having safe types and instead rely on enums themselves for any kind of communication to the machine. However I think it would be great for the next major version of `xstate` to allow specifying the states themselves instead of only relying on typegen. I know the "enum" support won't really be done and I am not asking for that, just maybe a generic, maybe even similar to `TStateSchema` so that we could specify what kind of states we have.

**Reply by Andarist** [reply]
I'll think about it - although I don't see a great value in this right now and supporting this would complicate our super complicated types even further.

> Unfortunately I would not really like disconnecting state names in the machine from the ones that we'd use during state machine "consumption". We have lots of people working on the project and it could get really messy really quickly having to keep that.

Could you give a code example of that? I would like to ensure that I understand your problem deeply to give it a proper thought and consideration.

**Reply by joffarex** [reply]
EDIT: btw, thanks a lot for actually being interested in my specific use case and thinking about possible improvements in the library itself as well. Not many framework contributors do the same. ❤️ 

Sure yeah, I am not asking to support it right away, so no worries there. 

My usecase would be like this: Today we have multiple enums as I have mentioned. All of them determine how the "process" which we are doing works and how different parts of the data gets treated. 

For example, we have "ProcessType", "Tenant", "Source" enums, and also one additional which determines on what step exactly the process is on. These are the major "pillars" or enums.

Lots of them have complicated names underneath, basically enum values are just abstracting away some abbreviations, standards, etc, as some of those gets sent downstream and then other systems use it for their own analytics, data filtering, etc.

The way we build the machine using `xstate` was to have small state machines based on each kind of config. So lets say for each "source" we have state machine, and also for each "ProcessStep" and "ProcessType" we have separate state machines, where all of them gets combined into one big machine that runs for each new "Process" that gets created. Right now that main "Process" has ~40-45 states, where majority starts other sub-machines that I had mentioned. 

Controlling this many strings, where lots of them follow specific standard not necessary for your average engineer to know + those enums coming from shared libraries so other microservies + UI can use them, would get too much work to, for example, hard code strings inside machines so we'd have type safety and also when trying to run this machine from specific "ProcessSteps" (Yes, the process can be paused and continued after user's input, for example) would become painful. 

This was my main use case why I decided to keep using enums, but try to maybe make machines type safe the way they are when you use typegen.

**Reply by Andarist** [reply]
I see your use case here - it's quite an advanced one! 😉 

1. part of my concern here is that this makes the bidirectional sync between the code and the Studio much harder to pull off. Are you interested in authoring your machines visually? How the process of renaming a state would look in this case? What would be displayed in the Studio as the state's name? Would it be the identifier that you have used? Or its value? Which exact parts of the code we'd change if you'd perform a rename?
2. Are you "just" interested in being able to use those enums~ as your state keys? Or do you actually want to enforce a schema through a generic (or similar), separately from the actual machine's definition?

**Reply by with-heart** [reply]
This is really interesting. Any way you could create a pared down reproduction @joffarex? I can't quite visualize it in my mind but would love to understand the system you're describing

**Reply by joffarex** [reply]
So to start off, I'll start answering @Andarist .

1. When I started using xstate, I really wanted to use the visual representation of it, however machines grew quickly so for me and my usecase it was much easier to, say, build some "base" in draw.io or something instead of relying on the xstate's visualization, so it is not a concern for me. 
2. The main thing that I wanted to have was to basically be able to use Enums as state names, which would be "typed" during `state.matches` or similar functionality. Not to allow passing into it, let's say some random string or something. I know its not a big deal per se if it accepts everything as of now, right, as we can always unit test the code and see if its incorrect, than it fails kind of thing. 

@with-heart I can try to visualize it for you, tho I don't think I'd have enough time at this moment to be able to give you code representation of it. If visualized machine with something like draw.io would be "enough".
